porch/controllers/rootsyncrollout/pkg/controllers/rootsyncrollout/controller.go

```diff
@@ -0,0 +1,317 @@
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	myFinalizerName := "config.porch.kpt.dev/finalizer"
+	if rootsyncrollout.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer and update the object. This is equivalent
+		// registering our finalizer.
+		if !controllerutil.ContainsFinalizer(&rootsyncrollout, myFinalizerName) {
+			controllerutil.AddFinalizer(&rootsyncrollout, myFinalizerName)
+			if err := r.Update(ctx, &rootsyncrollout); err != nil {
+				return ctrl.Result{}, fmt.Errorf("error adding finalizer: %w", err)
+			}
+		}
+	} else {
+		// The object is being deleted
+		if controllerutil.ContainsFinalizer(&rootsyncrollout, myFinalizerName) {
+			// our finalizer is present, so lets handle any external dependency
+			if err := r.deleteExternalResources(ctx, &rootsyncrollout); err != nil {
+				// if fail to delete the external dependency here, return with error
+				// so that it can be retried
+				return ctrl.Result{}, fmt.Errorf("have problem to delete external resource: %w", err)
+			}
+			// remove our finalizer from the list and update it.
+			controllerutil.RemoveFinalizer(&rootsyncrollout, myFinalizerName)
+			if err := r.Update(ctx, &rootsyncrollout); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to update %s after delete finalizer: %w", req.Name, err)
+			}
+		}
+		// Stop reconciliation as the item is being deleted
+		return ctrl.Result{}, nil
+	}
+
+	// Get all packages targeted by the Package selector
+	packageMap, err := r.getPackages(ctx, &rootsyncrollout)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	klog.Infof("Found %d packages: %v", len(packageMap), packageMap)
+
+	clusterList, err := r.getClusters(ctx, &rootsyncrollout)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	klog.Infof("Found %d clusters: %v", len(clusterList), clusterList)
+
+	for pkg := range packageMap {
+		revisions := packageMap[pkg]
+		var latest *porchapi.PackageRevision
+		for i := range revisions {
+			if isLatest(revisions[i]) {
+				l := *revisions[i]
+				latest = &l
+			}
+		}
+
+		for _, cluster := range clusterList {
+			klog.Infof("Checking for cluster %s and packagerevision %s (package %s)", cluster.GetName(), latest.Name, latest.Spec.PackageName)
+			repo, err := r.getRepository(ctx, latest.Spec.RepositoryName, latest.Namespace)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			rootSyncSetName := fmt.Sprintf("%s-%s", latest.Spec.PackageName, cluster.GetName())
+			rootSyncSetNamespace := req.Namespace
+
+			var rootsyncset rss.RootSyncSet
+			err = r.Get(ctx, types.NamespacedName{
+				Name:      rootSyncSetName,
+				Namespace: rootSyncSetNamespace,
+			}, &rootsyncset)
+
+			if err != nil && !apierrors.IsNotFound(err) {
+				return ctrl.Result{}, err
+			}
+
+			if apierrors.IsNotFound(err) {
+				klog.Infof("Creating rootsyncset %s", rootSyncSetName)
```

Comment:
I wonder if we can express this as creating a set of "child objects", and then use applylib or similar.  Should be less code and take care of pruning.  Not a blocker though!

Reply:
Yeah, I did consider that approach, but in the end decided on the more standard controller approach.